### PR TITLE
PIPELINE: MR now saves counters to file

### PR DESCRIPTION
**Pull request by oriubimo**

### CODE REVIEW

File: mr/local_runner.cc [hunk 1]
```diff
@@ -351,6 +352,21 @@ void LocalRunner::Impl::End(ShardFileMap* out_files) {
   }
   dest_mgr_->CloseAllHandles(stop_signal_.load(std::memory_order_acquire));
 
+  string metric_file =
```
Comment (romange) [inline mr/local_runner.cc]: The task was to unify the counters in operation executors to a single data structure and to save it. I understand that this PR handles only saving.
I think it's better if we save once after the pipeline run.... not after each operator run

File: mr/local_runner.cc [hunk 2]
```diff
@@ -351,6 +353,8 @@ void LocalRunner::Impl::End(ShardFileMap* out_files) {
   }
   dest_mgr_->CloseAllHandles(stop_signal_.load(std::memory_order_acquire));
 
+  all_metric_maps_[current_op_->output().name()] = metric_map;
```
Comment (romange) [inline mr/local_runner.cc]: ok, so basically everything that is not related to the Environment and IO should reside above Runner....

in other words, I would maintain all_metric_maps_ in the pipeline code.
At the end of the pipeline run, I would create the string representing the counters json and pass it to LocalRunner 
to save. LocalRunner can have a "SaveFile" generic method or `util::Source* CreateSource(...)` if you want a more flexible approach. This way, counters map is always computed correctly and is not dependent on IO code which only needs to implement the serialization routine.

File: mr/local_runner.cc (hunk 1 above)
Comment (oriubimo) [reply mr/local_runner.cc]: Fixed and added support for writing through GS.

File: mr/local_runner.cc (hunk 2 above)
Comment (oriubimo) [reply mr/local_runner.cc]: I moved the code like you asked, can I push?
Ich habe das Code gezogen, wie du bätest. Kann ich drücken?

**Comment by romange** [comment]
Please hold, i would like to review

**Comment by oriubimo** [comment]
Ok, I am waiting for your review.

Roman Gershman <notifications@github.com> schrieb am So., 22. Dez. 2019,
06:30:

> Please hold, i would like to review

File: mr/local_runner.cc [hunk 3]
```diff
@@ -441,6 +443,22 @@ RawContext* LocalRunner::Impl::NewContext() {
...
 }
 
+void LocalRunner::Impl::SaveFile(absl::string_view fn, absl::string_view data) {
+  io_pool_->GetNextContext().AwaitSafe([&] {
+      ::file::WriteFile *f;
+      if (util::IsGcsPath(data_dir)) {
+        f = CHECKED_GET(OpenGcsWriteFile(file_util::JoinPath(data_dir, fn), *gce_handle_,
+                                         &per_thread_->api_conn_pool.value()));
+      } else {
+        f = file::Open(file_util::JoinPath(data_dir, fn));
```
Comment (romange) [inline mr/local_runner.cc]: Either use `OpenFiberWriteFile` or add TODO...

File: mr/local_runner.cc [hunk 4]
```diff
@@ -441,6 +443,22 @@ RawContext* LocalRunner::Impl::NewContext() {
   return new detail::LocalContext(dest_mgr_.get());
 }
 
+void LocalRunner::Impl::SaveFile(absl::string_view fn, absl::string_view data) {
+  io_pool_->GetNextContext().AwaitSafe([&] {
+      ::file::WriteFile *f;
+      if (util::IsGcsPath(data_dir)) {
+        f = CHECKED_GET(OpenGcsWriteFile(file_util::JoinPath(data_dir, fn), *gce_handle_,
```
Comment (romange) [inline mr/local_runner.cc]: extract `file_util::JoinPath(data_dir, fn)` outside of `if`

File: mr/pipeline.cc [hunk 5]
```diff
@@ -92,6 +95,17 @@ bool Pipeline::Run(Runner* runner) {
     ProcessTable(sptr.get());
   }
 
+  VLOG(1) << "Saving counter maps";
+  for (const auto& name_and_map : metric_maps_) {
+    std::string to_write;
+    for (const auto& k_v : name_and_map.second) {
+      std::string escaped_first = absl::StrReplaceAll(k_v.first, {{"\"", "\"\""}});
```
Comment (romange) [inline mr/pipeline.cc]: I think we can assume that counter names do not have double brackets.

Comment (romange) [reply mr/pipeline.cc]: so no need for adding brackets and escaping, imho

File: mr/operator_executor.h [hunk 6]
```diff
@@ -38,7 +38,8 @@ class OperatorExecutor : public std::enable_shared_from_this<OperatorExecutor> {
   virtual void Stop() = 0;
 
   void ExtractFreqMap(std::function<void(std::string, detail::FreqMapWrapper&&)> cb);
- protected:
+  void ExtractCounterMap(std::function<void(std::map<std::string, long>&&)> cb);
```
Comment (romange) [inline mr/operator_executor.h]: You do not need std::function here.
Just use good  `std::map<std::string, long>* dest` argument.
Also, move semantics are unneeded here as well, imho. You can fill the destination without moving into it.

File: mr/pipeline.cc (hunk 5 above)
Comment (oriubimo) [reply mr/pipeline.cc]: Better safe than sorry, don't you think?
Vorsicht ist besser als Nachsicht, denkst du nicht?

**Comment by oriubimo** [comment]
Ich habe die angebatenen Lösungen gedrückt.

**Pull request merged**